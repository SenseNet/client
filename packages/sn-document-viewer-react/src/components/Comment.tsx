import Avatar from '@material-ui/core/Avatar'
import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import CardHeader from '@material-ui/core/CardHeader'
import Collapse from '@material-ui/core/Collapse'
import React, { useState } from 'react'
import { FunctionComponent } from 'react'
import { connect } from 'react-redux'
import styled from 'styled-components'
import { Comment as CommentType } from '../models/Comment'
import { componentType } from '../services'
import { deleteComment, RootReducerType } from '../store'

const mapStateToProps = (state: RootReducerType) => ({
  localization: state.sensenetDocumentViewer.localization,
})

const mapDispatchToProps = {
  deleteComment,
}

const StyledCardContent = styled(CardContent)`
  font-size: 14px;
  letter-spacing: 0.2px;
  line-height: 20px;
  word-wrap: break-word;
`
const DeleteButton = (props: CommentPropType) => (
  <Button size="small" onClick={() => props.deleteComment(props.id)}>
    {props.localization.delete || 'delete'}
  </Button>
)

/**
 * Comment prop type
 */
export type CommentPropType = componentType<typeof mapStateToProps, typeof mapDispatchToProps, CommentType>

const MAX_TEXT_LENGTH = 160

/**
 * Represents a single comment component.
 */
export const CommentComponent: FunctionComponent<CommentPropType> = props => {
  const isLongText = props.text && props.text.length > MAX_TEXT_LENGTH
  const [isOpen, setIsOpen] = useState(!isLongText)
  return (
    <Card>
      <CardHeader
        avatar={<Avatar src={props.createdBy.avatarUrl} alt={props.localization.avatarAlt} />}
        title={props.createdBy.displayName}
      />
      <Collapse in={isOpen} timeout="auto" collapsedHeight={isOpen ? '0px' : '78px'}>
        <StyledCardContent>{props.text}</StyledCardContent>
      </Collapse>
      <CardActions>
        {isLongText ? (
          <>
            <Button size="small" onClick={() => setIsOpen(!isOpen)}>
              {isOpen ? props.localization.showLess || 'Show less' : props.localization.showMore || 'Show more'}
            </Button>
            {isOpen ? <DeleteButton {...props} /> : null}
          </>
        ) : (
          <DeleteButton {...props} />
        )}
      </CardActions>
    </Card>
  )
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CommentComponent)
