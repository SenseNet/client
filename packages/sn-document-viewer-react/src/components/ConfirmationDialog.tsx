import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogTitle from '@material-ui/core/DialogTitle'
import React from 'react'
import { FunctionComponent } from 'react'

/**
 * Interface for confirmation dialog component properties
 */
export interface ConfirmationDialogProps {
  dialogTitle: string
  okButtonText?: string
  cancelButtonText?: string
  isOpen: boolean
  onClose: (isCanceled: boolean) => void
}

/**
 * Represents a confirmation dialog
 */
export const ConfirmationDialog: FunctionComponent<ConfirmationDialogProps> = props => {
  const handleCancel = () => props.onClose(true)
  const handleOk = () => props.onClose(false)
  return (
    <Dialog
      disableBackdropClick={true}
      disableEscapeKeyDown={true}
      maxWidth="md"
      aria-labelledby="confirmation-dialog-title"
      aria-describedby="confirmation-dialog-description"
      open={props.isOpen}>
      <DialogTitle id="confirmation-dialog-title">{props.dialogTitle}</DialogTitle>
      <DialogContent>
        <DialogContentText
          dangerouslySetInnerHTML={{ __html: props.children as string }}
          id="confirmation-dialog-description"
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={handleOk} variant="contained" color="primary">
          {props.okButtonText || 'ok'}
        </Button>
        <Button onClick={handleCancel}>{props.cancelButtonText || 'cancel'}</Button>
      </DialogActions>
    </Dialog>
  )
}
