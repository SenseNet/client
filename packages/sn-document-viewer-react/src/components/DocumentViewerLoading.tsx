import Typography from '@material-ui/core/Typography'
import React from 'react'
import { useLocalization } from '../hooks'

interface DocumentViewerLoadingProps {
  image: string
}

export const DocumentViewerLoading: React.FC<DocumentViewerLoadingProps> = (props) => {
  const localization = useLocalization()

  return (
    <div
      style={{
        display: 'flex',
        justifyContent: 'space-around',
        alignContent: 'center',
        alignItems: 'center',
        height: '100%',
      }}>
      <div
        style={{
          display: 'flex',
          justifyContent: 'space-between',
          alignItems: 'center',
          flexDirection: 'column',
          maxWidth: 500,
          margin: '.5em 0 .6em 0',
        }}>
        <img src={props.image} alt="Loader" />
        <Typography
          variant="h5"
          color="textSecondary"
          align="center"
          style={{ marginTop: '1rem', fontWeight: 'bolder' }}>
          {localization.loadingDocument}
        </Typography>
      </div>
    </div>
  )
}
