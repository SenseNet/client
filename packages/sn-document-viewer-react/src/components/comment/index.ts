export * from './comment'
export * from './create-comment'
export * from './delete-comment-button'
export * from './style'
