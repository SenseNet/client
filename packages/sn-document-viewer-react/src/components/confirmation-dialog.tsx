import { Button, Dialog, DialogActions, DialogContent, DialogContentText, DialogTitle } from '@material-ui/core'
import React from 'react'

/**
 * Interface for confirmation dialog component properties
 */
export interface ConfirmationDialogProps {
  dialogTitle: string
  okButtonText?: string
  cancelButtonText?: string
  isOpen: boolean
  onClose: (isCanceled: boolean) => void
}

/**
 * Represents a confirmation dialog
 */
export const ConfirmationDialog: React.FC<ConfirmationDialogProps> = (props) => {
  const handleCancel = () => props.onClose(true)
  const handleOk = () => props.onClose(false)
  return (
    <Dialog
      disableBackdropClick={true}
      disableEscapeKeyDown={true}
      maxWidth="md"
      aria-labelledby="confirmation-dialog-title"
      aria-describedby="confirmation-dialog-description"
      open={props.isOpen}>
      <DialogTitle id="confirmation-dialog-title">{props.dialogTitle}</DialogTitle>
      <DialogContent>
        <DialogContentText
          dangerouslySetInnerHTML={{ __html: props.children as string }}
          id="confirmation-dialog-description"
        />
      </DialogContent>
      <DialogActions>
        <Button onClick={handleOk} variant="contained" color="primary">
          {props.okButtonText || 'ok'}
        </Button>
        <Button onClick={handleCancel}>{props.cancelButtonText || 'cancel'}</Button>
      </DialogActions>
    </Dialog>
  )
}
