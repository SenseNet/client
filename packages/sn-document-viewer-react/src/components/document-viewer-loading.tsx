import { createStyles, makeStyles } from '@material-ui/core'
import Typography from '@material-ui/core/Typography'
import React from 'react'
import { useLocalization } from '../hooks'

const useStyles = makeStyles(() => {
  return createStyles({
    mainWrapper: {
      display: 'flex',
      justifyContent: 'space-around',
      alignContent: 'center',
      alignItems: 'center',
      height: '100%',
    },
    loaderWrapper: {
      display: 'flex',
      justifyContent: 'space-between',
      alignItems: 'center',
      flexDirection: 'column',
      maxWidth: 500,
      margin: '.5em 0 .6em 0',
    },
    typography: {
      marginTop: '1rem',
      fontWeight: 'bolder',
    },
  })
})

interface DocumentViewerLoadingProps {
  image: string
}

export const DocumentViewerLoading: React.FC<DocumentViewerLoadingProps> = (props) => {
  const classes = useStyles()
  const localization = useLocalization()

  return (
    <div className={classes.mainWrapper}>
      <div className={classes.loaderWrapper}>
        <img src={props.image} alt="Loader" />
        <Typography variant="h5" color="textSecondary" align="center" className={classes.typography}>
          {localization.loadingDocument}
        </Typography>
      </div>
    </div>
  )
}
