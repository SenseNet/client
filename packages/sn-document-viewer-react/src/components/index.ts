export * from './comment'
export * from './document-widgets'
export * from './page-widgets'
export * from './ConfirmationDialog'
export * from './DocumentViewer'
export * from './DocumentViewerError'
export * from './DocumentViewerLayout'
export * from './DocumentViewerLoading'
export * from './DocumentViewerRegeneratePreviews'
export * from './LayoutAppBar'
export * from './Page'
export * from './PageList'
export * from './styles'
