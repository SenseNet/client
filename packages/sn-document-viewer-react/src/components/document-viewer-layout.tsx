import Drawer from '@material-ui/core/Drawer'
import createStyles from '@material-ui/core/styles/createStyles'
import makeStyles from '@material-ui/core/styles/makeStyles'
import Typography from '@material-ui/core/Typography'
import React, { useCallback, useEffect } from 'react'
import { CommentsContext } from '../context/comments'
import { useLocalization, useViewerSettings, useViewerState } from '../hooks'
import { Comment, CreateComment } from './comment'
import { PageList, Thumbnails } from './'

export const THUMBNAIL_PADDING = 16
export const THUMBNAIL_NAME = 'Thumbnail'
export const THUMBNAIL_CONTAINER_ID = 'sn-document-viewer-thumbnails'
export const PAGE_PADDING = 8
export const PAGE_NAME = 'Page'
export const PAGE_CONTAINER_ID = 'sn-document-viewer-pages'

const useStyles = makeStyles(() => {
  return createStyles({
    mainWrapper: {
      width: '100%',
      height: '100%',
      display: 'flex',
      flexDirection: 'column',
      flexGrow: 1,
    },
    layout: {
      display: 'flex',
      flexGrow: 1,
      width: '100%',
      overflow: 'hidden',
      zIndex: 0,
      position: 'relative',
    },
    commentContainer: {
      display: 'flex',
      flexFlow: 'column',
      padding: '15px',
      justifyContent: 'space-between',
      alignItems: 'center',
      flexWrap: 'wrap',
      overflow: 'auto',
    },
  })
})

interface ScrollToOptions {
  containerId: string
  index: number
  itemName: string
  padding: number
  smoothScroll: boolean
}

export const DocumentViewerLayout: React.FC = () => {
  const classes = useStyles()
  const viewerState = useViewerState()
  const localization = useLocalization()
  const viewerSettings = useViewerSettings()

  const scrollToImage = useCallback(({ containerId, index, itemName, padding, smoothScroll }: ScrollToOptions) => {
    const container = document.getElementById(containerId)
    const item = document.querySelector(`.${itemName}`)
    if (container && container.scrollTo && item) {
      container.scrollTo({
        top: (item.clientHeight + padding * 4) * (index - 1),
        behavior: smoothScroll ? 'smooth' : 'auto',
      })
    }
  }, [])

  const scrollTo = useCallback(
    (index: number, smoothScroll = true) => {
      scrollToImage({
        containerId: PAGE_CONTAINER_ID,
        itemName: PAGE_NAME,
        padding: PAGE_PADDING,
        index,
        smoothScroll,
      })

      scrollToImage({
        containerId: THUMBNAIL_CONTAINER_ID,
        itemName: THUMBNAIL_NAME,
        padding: THUMBNAIL_PADDING,
        index,
        smoothScroll,
      })
    },
    [scrollToImage],
  )

  useEffect(() => {
    const observer = viewerState.pageToGo.subscribe(({ page }) => {
      scrollTo(page)
    })
    return () => observer.dispose()
  }, [scrollTo, viewerState.pageToGo])

  return (
    <div className={classes.mainWrapper}>
      {viewerSettings.renderAppBar()}
      <div className={classes.layout}>
        <Drawer
          variant={'persistent'}
          open={viewerState.showThumbnails}
          anchor="left"
          SlideProps={viewerSettings.drawerSlideProps}
          PaperProps={{
            style: {
              position: 'relative',
              width: viewerState.showThumbnails ? '200px' : 0,
              height: '100%',
              overflow: 'hidden',
            },
          }}>
          {viewerState.showThumbnails && <Thumbnails onPageClick={(index) => scrollTo(index)} />}
        </Drawer>
        <PageList onPageClick={(index) => scrollTo(index)} />
        <Drawer
          variant={'persistent'}
          open={viewerState.showComments}
          anchor="right"
          SlideProps={viewerSettings.drawerSlideProps}
          PaperProps={{
            style: {
              position: 'relative',
              width: viewerState.showComments ? '340px' : 0,
              height: '100%',
              overflow: 'hidden',
            },
          }}>
          <div className={classes.commentContainer}>
            <Typography variant="h4">{localization.commentSideBarTitle}</Typography>
            <CreateComment />
            <CommentsContext.Consumer>
              {(commentsContext) =>
                commentsContext.comments.map((comment) => <Comment key={comment.id} comment={comment} />)
              }
            </CommentsContext.Consumer>
          </div>
        </Drawer>
      </div>
    </div>
  )
}
