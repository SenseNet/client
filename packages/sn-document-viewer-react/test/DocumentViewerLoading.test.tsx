import React from 'react'
import { shallow } from 'enzyme'
import { DocumentViewerLoading } from '../src/components/DocumentViewerLoading'

describe('Document Viewer Loading component', () => {
  it('should render without crashing', () => {
    const wrapper = shallow(<DocumentViewerLoading image={''} />)
    expect(wrapper).toMatchSnapshot()
  })
})
