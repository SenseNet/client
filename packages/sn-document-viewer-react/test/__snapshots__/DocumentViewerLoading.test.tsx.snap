// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document Viewer Loading component should render without crashing 1`] = `
<div
  style={
    Object {
      "alignContent": "center",
      "alignItems": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "space-around",
    }
  }
>
  <Component
    style={
      Object {
        "position": "fixed",
        "top": 0,
      }
    }
  >
    <span />
  </Component>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
        "margin": ".5em 0 .6em 0",
        "maxWidth": 500,
      }
    }
  >
    <img
      src=""
    />
    <WithStyles(ForwardRef(Typography))
      align="center"
      color="textSecondary"
      style={
        Object {
          "fontWeight": "bolder",
        }
      }
      variant="h5"
    >
      Preview image generation is in progress
    </WithStyles(ForwardRef(Typography))>
  </div>
</div>
`;
